File: Algorithm/a_star.py
import heapq

class Node:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.g = float('inf')
        self.h = 0
        self.parent = None

    def __lt__(self, other):
        # Custom comparison method for heapq
        return (self.g + self.h) < (other.g + other.h)

def heuristic(node, goal):
    # Calculate the Manhattan distance as the heuristic
    return abs(node.x - goal.x) + abs(node.y - goal.y)

def get_neighbors(node, grid):
    # Returns the neighbors of a node in the grid
    neighbors = []
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]  # 4-directional movement
    for dx, dy in directions:
        nx, ny = node.x + dx, node.y + dy
        if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and not grid[nx][ny]:
            neighbors.append(Node(nx, ny))
    return neighbors

def reconstruct_path(node):
    # Reconstruct the path from the goal node to the start node
    path = []
    current = node
    while current is not None:
        path.append((current.x, current.y))
        current = current.parent
    return path[::-1]

def a_star(grid, start, goal):
    open_set = [start]
    start.g = 0
    start.h = heuristic(start, goal)

    while open_set:
        current = heapq.heappop(open_set)

        if current.x == goal.x and current.y == goal.y:
            return reconstruct_path(current)

        for neighbor in get_neighbors(current, grid):
            tentative_g = current.g + 1  # Assuming the cost of moving between adjacent nodes is 1
            if tentative_g < neighbor.g:
                neighbor.g = tentative_g
                neighbor.h = heuristic(neighbor, goal)
                neighbor.parent = current
                heapq.heappush(open_set, neighbor)

    return None


def print_grid(grid, col_digits=2):
    n_rows = len(grid)
    n_cols = len(grid[0])
    result = (
        ' ' * (col_digits + 1) + ''.join(str(i) for i in range(n_cols)) + '\n'
        + ' ' * (col_digits + 1) + ''.join('-' for i in range(n_cols)) + '\n')
    for i, row in enumerate(grid):
        result += (
            str(i).zfill(col_digits) + '|'
            + ''.join(str(elem) for elem in row) + '\n')
    return result


def draw_path(grid,path):
    for row, col in path:
        grid[row][col] = "\033[91m*\033[0m"  # ANSI escape code for red color

# Print the updated grid
    for row in grid:
        print(" ".join(str(cell) for cell in row))

if __name__ == "__main__":
    # Example usage
    grid = [
        [0, 0, 0, 0, 0],
        [0, 1, 1, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 1, 1, 0],
        [0, 0, 0, 0, 0]
    ]

    print(print_grid(grid))
    
    start_node = Node(0, 0)
    goal_node = Node(2, 3)

    path = a_star(grid, start_node, goal_node)

    sorted_path = [[x, y] for x, y in path]

    if path:
        print("Path found:", path)
        print("\n")
        draw_path(grid,sorted_path)

    else:
        print("Path not found.")

File: Algorithm/a_star.cpp
#include <algorithm>
#include <cmath>
#include <functional>
#include <iostream>
#include <queue>
#include <unordered_set>
#include <vector>

using namespace std;

class Node {
public:
  int x, y;
  float g = INFINITY;
  float h = 0;
  Node *parent = nullptr;

  Node(int x = 0, int y = 0) : x(x), y(y) {}

  bool operator<(const Node &other) const {
    return (g + h) > (other.g + other.h); // Use > for min-heap priority_queue
  }

  bool operator==(const Node &other) const {
    return x == other.x && y == other.y;
  }
};

// Custom hash function for Node to use in unordered_set
struct NodeHash {
  std::size_t operator()(const Node &n) const {
    return std::hash<int>()(n.x) ^ std::hash<int>()(n.y);
  }
};

float heuristic(const Node &node, const Node &goal) {
  return abs(node.x - goal.x) + abs(node.y - goal.y);
}

vector<Node> get_neighbors(const Node &node, const vector<vector<int>> &grid) {
  vector<Node> neighbors;
  int directions[4][2] = {
      {0, 1}, {0, -1}, {1, 0}, {-1, 0}}; // 4-directional movement

  for (const auto &dir : directions) {
    int nx = node.x + dir[0];
    int ny = node.y + dir[1];

    if (nx >= 0 && nx < grid.size() && ny >= 0 && ny < grid[0].size() &&
        grid[nx][ny] == 0) {
      neighbors.push_back({nx, ny});
    }
  }

  return neighbors;
}

vector<pair<int, int>> reconstruct_path(const Node *node) {
  vector<pair<int, int>> path;
  const Node *current = node;

  while (current != nullptr) {
    path.push_back({current->x, current->y});
    current = current->parent;
  }

  reverse(path.begin(), path.end());
  return path;
}

vector<pair<int, int>> a_star(const vector<vector<int>> &grid, Node start,
                              const Node &goal) {
  priority_queue<Node> open_set;
  unordered_set<Node, NodeHash> closed_set;

  start.g = 0;
  start.h = heuristic(start, goal);
  open_set.push(start);

  while (!open_set.empty()) {
    Node current = open_set.top();
    open_set.pop();

    if (current == goal) {
      return reconstruct_path(&current);
    }

    closed_set.insert(current);

    for (auto neighbor : get_neighbors(current, grid)) {
      if (closed_set.find(neighbor) != closed_set.end()) {
        continue;
      }

      float tentative_g =
          current.g + 1; // Assuming cost of moving between adjacent nodes is 1

      if (tentative_g < neighbor.g) {
        neighbor.g = tentative_g;
        neighbor.h = heuristic(neighbor, goal);
        neighbor.parent = new Node(current); // Use new to store parent

        open_set.push(neighbor);
      }
    }
  }

  return {}; // No path found
}

void draw_path(vector<vector<int>> &grid, const vector<pair<int, int>> &path) {
  for (const auto &point : path) {
    grid[point.first][point.second] = -1; // Marking the path
  }
}

int main() {
  vector<vector<int>> grid = {{0, 0, 0, 0, 0},
                              {0, 1, 1, 0, 0},
                              {0, 0, 0, 0, 0},
                              {0, 0, 1, 1, 0},
                              {0, 0, 0, 0, 0}};

  cout << "Initial grid:" << endl;
  for (const auto &row : grid) {
    for (int cell : row) {
      cout << cell << " ";
    }
    cout << endl;
  }

  Node start_node{0, 0};
  Node goal_node{3, 2}; // Adjust goal position to be within the grid

  vector<pair<int, int>> path = a_star(grid, start_node, goal_node);

  if (!path.empty()) {
    cout << "\nPath found:" << endl;
    draw_path(grid, path);

    for (const auto &row : grid) {
      for (int cell : row) {
        if (cell == -1) {
          cout << "* ";
        } else {
          cout << cell << " ";
        }
      }
      cout << endl;
    }
  } else {
    cout << "Path not found." << endl;
  }

  return 0;
}
